In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("world_economics.csv")
df 

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
0,Tunisia,Tunisian dinar,Tunis,Arabic,34.000000,9.000000,163610.0,Africa,Northern Africa,"['DZA', 'LBY']",53,1.80,7.50,5.30,15.70,-6.6,79.8,-2.7,11.89
1,Vietnam,Vietnamese đồng,Hanoi,Vietnamese,16.166667,107.833333,331212.0,Asia,South-Eastern Asia,"['KHM', 'CHN', 'LAO']",476,6.88,4.50,3.19,2.24,-4.0,32.9,6.1,101.30
2,Ecuador,United States dollar,Quito,Spanish,-2.000000,-77.500000,276841.0,Americas,South America,"['COL', 'PER']",125,3.46,8.50,0.72,4.50,-1.9,50.6,2.8,17.97
3,Puerto Rico,United States dollar,San Juan,English,18.250000,-66.500000,8870.0,Americas,Caribbean,NaN,126,NaN,NaN,1.20,5.70,NaN,NaN,NaN,3.20
4,United States,United States dollar,"Washington, D.C.",English,38.000000,-97.000000,9372610.0,Americas,North America,"['CAN', 'MEX']",29185,3.00,4.50,2.70,4.20,-6.4,124.3,-3.9,341.15
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
168,New Caledonia,CFP franc,Nouméa,French,-21.500000,165.500000,18575.0,Oceania,Melanesia,NaN,10,NaN,2.00,0.80,11.20,NaN,NaN,NaN,0.29
169,Kosovo,Euro,Pristina,Albanian,42.666667,21.166667,10908.0,Europe,Southeast Europe,"['ALB', 'MKD', 'MNE', 'SRB']",11,NaN,NaN,4.30,10.80,0.2,16.9,-9.0,1.60
170,Malaysia,Malaysian ringgit,Kuala Lumpur,English,2.500000,112.500000,330803.0,Asia,South-Eastern Asia,"['BRN', 'IDN', 'THA']",422,2.10,2.75,1.20,3.00,-4.1,70.4,1.7,34.10
171,Fiji,Fijian dollar,Suva,English,-17.713400,178.065000,18272.0,Oceania,Melanesia,NaN,6,NaN,0.25,-0.40,4.30,-3.4,78.3,-8.7,0.93


- **borders** → الدول اللي لها حدود مشتركة مع الدولة 
- **GDP** → الناتج المحلي الإجمالي بمليار دولار  
- **GDP Growth** → معدل نمو الناتج المحلي % (قد إيه الاقتصاد بينمو)  
- **Interest Rate** → سعر الفائدة % (اللي بيحدده البنك المركزي)  
- **Inflation Rate** → معدل التضخم % (زيادة أسعار السلع والخدمات)  
- **Gov. Budget** → نسبة العجز أو الفائض في ميزانية الحكومة من الناتج المحلي % (سالب = عجز)  
- **Debt/GDP** → نسبة الدين العام للناتج المحلي % (مدى اعتماد الدولة على الديون)  
- **Current Account** → الحساب الجاري % من الناتج المحلي (يعبر عن الفرق بين الصادرات والواردات + التحويلات)  

In [2]:
df.columns

Index(['name', 'currency', 'capital', 'languages', 'latitude', 'longitude',
       'area', 'region', 'subregion', 'borders', 'GDP', 'GDP Growth',
       'Interest Rate', 'Inflation Rate', 'Jobless Rate', 'Gov. Budget',
       'Debt/GDP', 'Current Account', 'Population'],
      dtype='object')

In [3]:
df.shape

(173, 19)

In [4]:
df.head()

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
0,Tunisia,Tunisian dinar,Tunis,Arabic,34.000000,9.000000,163610.0,Africa,Northern Africa,"['DZA', 'LBY']",53,1.80,7.5,5.30,15.70,-6.6,79.8,-2.7,11.89
1,Vietnam,Vietnamese đồng,Hanoi,Vietnamese,16.166667,107.833333,331212.0,Asia,South-Eastern Asia,"['KHM', 'CHN', 'LAO']",476,6.88,4.5,3.19,2.24,-4.0,32.9,6.1,101.30
2,Ecuador,United States dollar,Quito,Spanish,-2.000000,-77.500000,276841.0,Americas,South America,"['COL', 'PER']",125,3.46,8.5,0.72,4.50,-1.9,50.6,2.8,17.97
3,Puerto Rico,United States dollar,San Juan,English,18.250000,-66.500000,8870.0,Americas,Caribbean,NaN,126,NaN,NaN,1.20,5.70,NaN,NaN,NaN,3.20
4,United States,United States dollar,"Washington, D.C.",English,38.000000,-97.000000,9372610.0,Americas,North America,"['CAN', 'MEX']",29185,3.00,4.5,2.70,4.20,-6.4,124.3,-3.9,341.15


In [5]:
df.tail()

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
168,New Caledonia,CFP franc,Nouméa,French,-21.500000,165.500000,18575.0,Oceania,Melanesia,NaN,10,NaN,2.00,0.8,11.2,NaN,NaN,NaN,0.29
169,Kosovo,Euro,Pristina,Albanian,42.666667,21.166667,10908.0,Europe,Southeast Europe,"['ALB', 'MKD', 'MNE', 'SRB']",11,NaN,NaN,4.3,10.8,0.2,16.9,-9.0,1.60
170,Malaysia,Malaysian ringgit,Kuala Lumpur,English,2.500000,112.500000,330803.0,Asia,South-Eastern Asia,"['BRN', 'IDN', 'THA']",422,2.1,2.75,1.2,3.0,-4.1,70.4,1.7,34.10
171,Fiji,Fijian dollar,Suva,English,-17.713400,178.065000,18272.0,Oceania,Melanesia,NaN,6,NaN,0.25,-0.4,4.3,-3.4,78.3,-8.7,0.93
172,Jamaica,Jamaican dollar,Kingston,English,18.250000,-77.500000,10991.0,Americas,Caribbean,NaN,20,0.8,5.75,3.3,3.3,-1.0,67.9,1.3,2.82


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173 entries, 0 to 172
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             173 non-null    object 
 1   currency         173 non-null    object 
 2   capital          172 non-null    object 
 3   languages        173 non-null    object 
 4   latitude         173 non-null    float64
 5   longitude        173 non-null    float64
 6   area             173 non-null    float64
 7   region           173 non-null    object 
 8   subregion        173 non-null    object 
 9   borders          149 non-null    object 
 10  GDP              173 non-null    int64  
 11  GDP Growth       101 non-null    float64
 12  Interest Rate    162 non-null    float64
 13  Inflation Rate   172 non-null    float64
 14  Jobless Rate     171 non-null    float64
 15  Gov. Budget      166 non-null    float64
 16  Debt/GDP         165 non-null    float64
 17  Current Account 

In [7]:
df.describe()

,latitude,longitude,area,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
count,173.000000,173.000000,1.730000e+02,173.000000,101.000000,162.000000,172.000000,171.000000,166.000000,165.000000,167.000000,173.000000
mean,20.437142,22.371774,7.561437e+05,637.508671,1.115644,7.802222,8.348663,7.363099,-2.822831,60.945879,-0.545269,45.570058
std,23.961270,58.849734,2.007481e+06,2703.930867,3.820764,8.169314,21.367006,6.065999,4.265735,38.998840,8.871213,155.896870
min,-41.000000,-102.000000,2.020000e+00,2.000000,-23.700000,0.000000,-9.100000,0.100000,-17.600000,2.300000,-31.500000,0.040000
25%,6.500000,-2.000000,3.839400e+04,20.000000,0.100000,2.812500,1.275000,3.200000,-4.975000,36.100000,-4.900000,3.540000
50%,20.000000,24.000000,1.810350e+05,72.000000,0.700000,5.250000,3.350000,5.200000,-3.175000,54.900000,-1.000000,10.580000
75%,40.000000,49.000000,5.870410e+05,373.000000,1.800000,9.187500,5.300000,9.115000,-1.100000,76.500000,3.210000,34.430000
max,65.000000,178.065000,1.709824e+07,29185.000000,18.600000,59.230000,172.000000,33.200000,21.600000,272.000000,31.400000,1408.000000


In [8]:
# outlier of GDP Growth column
Q3 = 1.800000
IQR = 1.800000 - 0.100000
MAX_UP = Q3 + 1.5 * IQR
MAX_UP

4.35

In [9]:
df[df['GDP Growth'] > MAX_UP]

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
1,Vietnam,Vietnamese đồng,Hanoi,Vietnamese,16.166667,107.833333,331212.0,Asia,South-Eastern Asia,"['KHM', 'CHN', 'LAO']",476,6.88,4.50,3.19,2.24,-4.0,32.9,6.1,101.30
44,Mongolia,Mongolian tögrög,Ulan Bator,Mongolian,46.000000,105.000000,1564110.0,Asia,Eastern Asia,"['CHN', 'RUS']",24,6.10,12.00,8.10,4.50,0.2,33.0,-6.9,3.54
57,Algeria,Algerian dinar,Algiers,Arabic,28.000000,3.000000,2381741.0,Africa,Northern Africa,"['TUN', 'LBY', 'NER', 'ESH', 'MRT', 'MLI', 'MAR']",264,18.60,3.00,-0.22,11.43,-9.3,46.2,1.3,46.81
62,Bolivia,Bolivian boliviano,Sucre,Aymara,-17.000000,-65.000000,1098581.0,Americas,South America,"['ARG', 'BRA', 'CHL', 'PRY', 'PER']",50,7.89,4.05,24.86,2.40,-7.8,95.0,-2.7,12.41
77,Bangladesh,Bangladeshi taka,Dhaka,Bengali,24.000000,90.000000,147570.0,Asia,Southern Asia,"['MMR', 'IND']",450,5.82,10.00,8.55,4.70,-4.6,31.7,-1.4,171.60
104,Nigeria,Nigerian naira,Abuja,English,10.000000,8.000000,923768.0,Africa,Western Africa,"['BEN', 'CMR', 'TCD', 'NER']",188,10.00,27.50,21.88,4.30,-6.1,52.9,-0.5,232.68
132,Pakistan,Pakistani rupee,Islamabad,English,30.000000,70.000000,881912.0,Asia,Southern Asia,"['AFG', 'CHN', 'IND', 'IRN']",373,5.79,11.00,4.10,5.50,-6.8,80.0,-0.5,251.27
159,Egypt,Egyptian pound,Cairo,Arabic,27.000000,30.000000,1002450.0,Africa,Northern Africa,"['ISR', 'LBY', 'PSE', 'SDN']",389,9.80,24.00,13.90,6.10,-7.1,82.9,-5.3,107.20
160,Kuwait,Kuwaiti dinar,Kuwait City,Arabic,29.500000,45.750000,17818.0,Asia,Western Asia,"['IRQ', 'SAU']",160,4.70,4.00,2.39,2.10,21.6,3.0,28.2,4.79


In [10]:
df.duplicated().sum()

0

In [11]:
df.isna().sum()

name                0
currency            0
capital             1
languages           0
latitude            0
longitude           0
area                0
region              0
subregion           0
borders            24
GDP                 0
GDP Growth         72
Interest Rate      11
Inflation Rate      1
Jobless Rate        2
Gov. Budget         7
Debt/GDP            8
Current Account     6
Population          0
dtype: int64

In [12]:
df[['region','subregion','GDP','area']]

,region,subregion,GDP,area
0,Africa,Northern Africa,53,163610.0
1,Asia,South-Eastern Asia,476,331212.0
2,Americas,South America,125,276841.0
3,Americas,Caribbean,126,8870.0
4,Americas,North America,29185,9372610.0
...,...,...,...,...
168,Oceania,Melanesia,10,18575.0
169,Europe,Southeast Europe,11,10908.0
170,Asia,South-Eastern Asia,422,330803.0
171,Oceania,Melanesia,6,18272.0


In [13]:
# clarify all regions in dataset
df['region'].value_counts()

region
Africa      50
Asia        48
Europe      41
Americas    29
Oceania      5
Name: count, dtype: int64

In [14]:
df['subregion'].value_counts()

subregion
Western Asia                 17
Eastern Africa               16
Western Africa               15
South America                12
Northern Europe              10
South-Eastern Asia           10
Southern Asia                 9
Western Europe                8
Caribbean                     8
Southeast Europe              8
Middle Africa                 8
Eastern Asia                  7
Northern Africa               6
Central America               6
Southern Europe               6
Central Asia                  5
Southern Africa               5
Central Europe                5
Eastern Europe                4
North America                 3
Melanesia                     3
Australia and New Zealand     2
Name: count, dtype: int64

In [15]:
df.head()

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
0,Tunisia,Tunisian dinar,Tunis,Arabic,34.000000,9.000000,163610.0,Africa,Northern Africa,"['DZA', 'LBY']",53,1.80,7.5,5.30,15.70,-6.6,79.8,-2.7,11.89
1,Vietnam,Vietnamese đồng,Hanoi,Vietnamese,16.166667,107.833333,331212.0,Asia,South-Eastern Asia,"['KHM', 'CHN', 'LAO']",476,6.88,4.5,3.19,2.24,-4.0,32.9,6.1,101.30
2,Ecuador,United States dollar,Quito,Spanish,-2.000000,-77.500000,276841.0,Americas,South America,"['COL', 'PER']",125,3.46,8.5,0.72,4.50,-1.9,50.6,2.8,17.97
3,Puerto Rico,United States dollar,San Juan,English,18.250000,-66.500000,8870.0,Americas,Caribbean,NaN,126,NaN,NaN,1.20,5.70,NaN,NaN,NaN,3.20
4,United States,United States dollar,"Washington, D.C.",English,38.000000,-97.000000,9372610.0,Americas,North America,"['CAN', 'MEX']",29185,3.00,4.5,2.70,4.20,-6.4,124.3,-3.9,341.15


In [16]:
df.groupby("region")['GDP'].sum()

region
Africa       2749
Americas    38408
Asia        40253
Europe      26818
Oceania      2061
Name: GDP, dtype: int64

In [17]:
# clarify total area of all capital
df['area'].sum()

130812865.02

In [65]:
df.groupby(["region", "subregion"])['area'].sum()

region    subregion                
Africa    Eastern Africa                5968014.00
          Middle Africa                 4886590.00
          Northern Africa               7639959.00
          Southern Africa               3049764.00
          Western Africa                6105440.00
Americas  Caribbean                      225503.00
          Central America                499312.00
          North America                21321655.00
          South America                17737675.00
Asia      Central Asia                  4003451.00
          Eastern Asia                 11786538.00
          South-Eastern Asia            4522945.00
          Southern Asia                 6868982.00
          Western Asia                  4820813.00
Europe    Central Europe                 558888.00
          Eastern Europe               17943188.00
          Northern Europe               1746874.00
          Southeast Europe               598902.00
          Southern Europe               104097

In [21]:
# handling missing value in borders, capital columns
df['borders'].replace({np.nan: "[]"}, inplace = True)
df['capital'].replace({np.nan: "UnKnown"}, inplace = True)

In [22]:
# this capital which has null value ------> Unknown
df.loc[166,:]

name                         Macau
currency           Macanese pataca
capital                    UnKnown
languages               Portuguese
latitude                 22.166667
longitude                   113.55
area                          30.0
region                        Asia
subregion             Eastern Asia
borders                    ['CHN']
GDP                             50
GDP Growth                     NaN
Interest Rate                 4.75
Inflation Rate                0.12
Jobless Rate                   1.9
Gov. Budget                    NaN
Debt/GDP                       NaN
Current Account               31.4
Population                    0.69
Name: 166, dtype: object

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173 entries, 0 to 172
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             173 non-null    object 
 1   currency         173 non-null    object 
 2   capital          173 non-null    object 
 3   languages        173 non-null    object 
 4   latitude         173 non-null    float64
 5   longitude        173 non-null    float64
 6   area             173 non-null    float64
 7   region           173 non-null    object 
 8   subregion        173 non-null    object 
 9   borders          173 non-null    object 
 10  GDP              173 non-null    int64  
 11  GDP Growth       101 non-null    float64
 12  Interest Rate    162 non-null    float64
 13  Inflation Rate   172 non-null    float64
 14  Jobless Rate     171 non-null    float64
 15  Gov. Budget      166 non-null    float64
 16  Debt/GDP         165 non-null    float64
 17  Current Account 

In [24]:
df.head()

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
0,Tunisia,Tunisian dinar,Tunis,Arabic,34.000000,9.000000,163610.0,Africa,Northern Africa,"['DZA', 'LBY']",53,1.80,7.5,5.30,15.70,-6.6,79.8,-2.7,11.89
1,Vietnam,Vietnamese đồng,Hanoi,Vietnamese,16.166667,107.833333,331212.0,Asia,South-Eastern Asia,"['KHM', 'CHN', 'LAO']",476,6.88,4.5,3.19,2.24,-4.0,32.9,6.1,101.30
2,Ecuador,United States dollar,Quito,Spanish,-2.000000,-77.500000,276841.0,Americas,South America,"['COL', 'PER']",125,3.46,8.5,0.72,4.50,-1.9,50.6,2.8,17.97
3,Puerto Rico,United States dollar,San Juan,English,18.250000,-66.500000,8870.0,Americas,Caribbean,[],126,NaN,NaN,1.20,5.70,NaN,NaN,NaN,3.20
4,United States,United States dollar,"Washington, D.C.",English,38.000000,-97.000000,9372610.0,Americas,North America,"['CAN', 'MEX']",29185,3.00,4.5,2.70,4.20,-6.4,124.3,-3.9,341.15


In [25]:
# handling missing values in GDP Growth column
df["GDP Growth"] = df["GDP Growth"].fillna(df.groupby("region")["GDP Growth"].transform("median"))

In [26]:
df[['region','subregion','GDP Growth']]

,region,subregion,GDP Growth
0,Africa,Northern Africa,1.80
1,Asia,South-Eastern Asia,6.88
2,Americas,South America,3.46
3,Americas,Caribbean,0.60
4,Americas,North America,3.00
...,...,...,...
168,Oceania,Melanesia,0.50
169,Europe,Southeast Europe,0.25
170,Asia,South-Eastern Asia,2.10
171,Oceania,Melanesia,0.50


In [66]:
df[["capital", "subregion","GDP Growth"]]

,capital,subregion,GDP Growth
0,Tunis,Northern Africa,1.80
1,Hanoi,South-Eastern Asia,6.88
2,Quito,South America,3.46
3,San Juan,Caribbean,0.60
4,"Washington, D.C.",North America,3.00
...,...,...,...
168,Nouméa,Melanesia,0.50
169,Pristina,Southeast Europe,0.25
170,Kuala Lumpur,South-Eastern Asia,2.10
171,Suva,Melanesia,0.50


In [28]:
# languages column
df.groupby('languages')['GDP'].mean().sort_values(ascending = False)

languages
Chinese               9750.5
Japanese              4026.0
Italian               2373.0
Korean                1713.0
Indonesian            1396.0
                       ...  
Malay                   15.0
Montenegrin              8.0
Maldivian                7.0
Dzongkha                 3.0
Seychellois Creole       2.0
Name: GDP, Length: 61, dtype: float64

In [29]:
df['languages'].value_counts()

languages
English              37
Arabic               25
French               19
Spanish              15
Portuguese            6
                     ..
Norwegian Nynorsk     1
Lithuanian            1
Slovak                1
Swedish               1
Sinhala               1
Name: count, Length: 61, dtype: int64

In [30]:
df.head()

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
0,Tunisia,Tunisian dinar,Tunis,Arabic,34.000000,9.000000,163610.0,Africa,Northern Africa,"['DZA', 'LBY']",53,1.80,7.5,5.30,15.70,-6.6,79.8,-2.7,11.89
1,Vietnam,Vietnamese đồng,Hanoi,Vietnamese,16.166667,107.833333,331212.0,Asia,South-Eastern Asia,"['KHM', 'CHN', 'LAO']",476,6.88,4.5,3.19,2.24,-4.0,32.9,6.1,101.30
2,Ecuador,United States dollar,Quito,Spanish,-2.000000,-77.500000,276841.0,Americas,South America,"['COL', 'PER']",125,3.46,8.5,0.72,4.50,-1.9,50.6,2.8,17.97
3,Puerto Rico,United States dollar,San Juan,English,18.250000,-66.500000,8870.0,Americas,Caribbean,[],126,0.60,NaN,1.20,5.70,NaN,NaN,NaN,3.20
4,United States,United States dollar,"Washington, D.C.",English,38.000000,-97.000000,9372610.0,Americas,North America,"['CAN', 'MEX']",29185,3.00,4.5,2.70,4.20,-6.4,124.3,-3.9,341.15


In [31]:
# handling missing values in Interest Rate, Inflation Rate columns
df["Interest Rate"] = df["Interest Rate"].fillna(df.groupby("region")["Interest Rate"].transform("median"))
df["Inflation Rate"] = df["Inflation Rate"].fillna(df["Inflation Rate"].median())

In [32]:
# top 5 countries in Interest Rate 
df[['name','region','Interest Rate']].sort_values(by = 'Interest Rate',ascending = False).head()

,name,region,Interest Rate
120,Venezuela,Americas,59.23
33,Turkey,Asia,43.00
61,Zimbabwe,Africa,35.00
158,Argentina,Americas,29.00
17,Sudan,Africa,28.30


In [67]:
# top 5 countries in Inflation Rate 
df[['name','region','Inflation Rate']].sort_values(by = 'Inflation Rate',ascending = False).head()

,name,region,Inflation Rate
120,Venezuela,Americas,172.00
17,Sudan,Africa,146.60
42,South Sudan,Africa,112.60
61,Zimbabwe,Africa,95.80
137,Palestine,Asia,59.35


In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173 entries, 0 to 172
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             173 non-null    object 
 1   currency         173 non-null    object 
 2   capital          173 non-null    object 
 3   languages        173 non-null    object 
 4   latitude         173 non-null    float64
 5   longitude        173 non-null    float64
 6   area             173 non-null    float64
 7   region           173 non-null    object 
 8   subregion        173 non-null    object 
 9   borders          173 non-null    object 
 10  GDP              173 non-null    int64  
 11  GDP Growth       173 non-null    float64
 12  Interest Rate    173 non-null    float64
 13  Inflation Rate   173 non-null    float64
 14  Jobless Rate     171 non-null    float64
 15  Gov. Budget      166 non-null    float64
 16  Debt/GDP         165 non-null    float64
 17  Current Account 

In [36]:
# Jobless Rate column
df[['subregion','Jobless Rate']]

,subregion,Jobless Rate
0,Northern Africa,15.70
1,South-Eastern Asia,2.24
2,South America,4.50
3,Caribbean,5.70
4,North America,4.20
...,...,...
168,Melanesia,11.20
169,Southeast Europe,10.80
170,South-Eastern Asia,3.00
171,Melanesia,4.30


In [37]:
# handling missing values in Jobless Rate column
df["Jobless Rate"] = df["Jobless Rate"].fillna(df["Jobless Rate"].median())

In [68]:
# Countries have Jobless Rate greater than or equal 10%
df.loc[df['Jobless Rate'] >= 10, ['name','region','Jobless Rate']]

,name,region,Jobless Rate
0,Tunisia,Africa,15.70
9,Afghanistan,Asia,13.30
11,Angola,Africa,29.40
15,Nepal,Asia,10.70
16,Iraq,Asia,15.50
17,Sudan,Africa,20.80
18,Djibouti,Africa,25.90
19,Mozambique,Africa,25.04
20,South Africa,Africa,33.20
23,Rwanda,Africa,13.40


In [62]:
# top 5 countries in Jobless Rate 
df[['name','region','Jobless Rate']].sort_values(by = 'Jobless Rate',ascending = False).head()

,name,region,Jobless Rate
20,South Africa,Africa,33.20
11,Angola,Africa,29.40
137,Palestine,Asia,28.60
18,Djibouti,Africa,25.90
19,Mozambique,Africa,25.04


In [40]:
# Avg Jobless Rate in each region
df.groupby('region')['Jobless Rate'].sum()

region
Africa      515.54
Americas    173.40
Asia        303.33
Europe      249.62
Oceania      27.60
Name: Jobless Rate, dtype: float64

In [69]:
# Gov. Budget column 
df[['name','borders','Gov. Budget']]

,name,borders,Gov. Budget
0,Tunisia,"['DZA', 'LBY']",-6.600
1,Vietnam,"['KHM', 'CHN', 'LAO']",-4.000
2,Ecuador,"['COL', 'PER']",-1.900
3,Puerto Rico,[],-3.145
4,United States,"['CAN', 'MEX']",-6.400
...,...,...,...
168,New Caledonia,[],-3.250
169,Kosovo,"['ALB', 'MKD', 'MNE', 'SRB']",0.200
170,Malaysia,"['BRN', 'IDN', 'THA']",-4.100
171,Fiji,[],-3.400


In [42]:
# handling missing values in Gov. Budget column
df["Gov. Budget"] = df["Gov. Budget"].fillna(df.groupby("region")["Gov. Budget"].transform("median"))

In [70]:
# countries have a government budget deficit
deficit_countries = df[df["Gov. Budget"] < 0][["name", "Gov. Budget"]]

# countries have a government budget surplus 
surplus_countries = df[df["Gov. Budget"] > 0][["name", "Gov. Budget"]]
print(surplus_countries)

                      name  Gov. Budget
28              Kyrgyzstan          2.4
36    United Arab Emirates          4.5
38                  Greece          1.3
44                Mongolia          0.2
45                 Moldova          3.9
51   Republic of the Congo          2.7
63                 Lesotho          8.4
65            Turkmenistan          0.8
66                  Norway         13.2
78                  Taiwan          0.4
95              Luxembourg          1.0
101             Tajikistan          1.4
106                 Cyprus          4.3
107               Portugal          0.7
110                Denmark          4.5
118             Seychelles          0.1
129             Montenegro          0.2
130                  Libya          8.5
131                  Qatar          0.7
134                 Brunei          1.1
137              Palestine          0.5
144              Australia          0.6
146                Ireland          4.3
147              Nicaragua          0.3


In [44]:
# top 5 countries by government budget, including their name and region
df[['name','region','Gov. Budget']].sort_values(by = 'Gov. Budget',ascending = False).head()

,name,region,Gov. Budget
160,Kuwait,Asia,21.6
66,Norway,Europe,13.2
130,Libya,Africa,8.5
63,Lesotho,Africa,8.4
157,Haiti,Americas,7.1


In [63]:
df.groupby('region')['Gov. Budget'].mean()

region
Africa     -3.508000
Americas   -3.463621
Asia       -2.626875
Europe     -1.785366
Oceania    -2.710000
Name: Gov. Budget, dtype: float64

In [46]:
df.head()

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
0,Tunisia,Tunisian dinar,Tunis,Arabic,34.000000,9.000000,163610.0,Africa,Northern Africa,"['DZA', 'LBY']",53,1.80,7.50,5.30,15.70,-6.600,79.8,-2.7,11.89
1,Vietnam,Vietnamese đồng,Hanoi,Vietnamese,16.166667,107.833333,331212.0,Asia,South-Eastern Asia,"['KHM', 'CHN', 'LAO']",476,6.88,4.50,3.19,2.24,-4.000,32.9,6.1,101.30
2,Ecuador,United States dollar,Quito,Spanish,-2.000000,-77.500000,276841.0,Americas,South America,"['COL', 'PER']",125,3.46,8.50,0.72,4.50,-1.900,50.6,2.8,17.97
3,Puerto Rico,United States dollar,San Juan,English,18.250000,-66.500000,8870.0,Americas,Caribbean,[],126,0.60,5.75,1.20,5.70,-3.145,NaN,NaN,3.20
4,United States,United States dollar,"Washington, D.C.",English,38.000000,-97.000000,9372610.0,Americas,North America,"['CAN', 'MEX']",29185,3.00,4.50,2.70,4.20,-6.400,124.3,-3.9,341.15


In [47]:
# handling missing values in Debt/GDP , Current Account columns
df["Debt/GDP"] = df["Debt/GDP"].fillna(df.groupby("region")["Debt/GDP"].transform("median"))
df["Current Account"] = df["Current Account"].fillna(df.groupby("region")["Current Account"].transform("median"))

In [71]:
df[['name', 'Debt/GDP']].sort_values(by = 'Debt/GDP', ascending = False)

,name,Debt/GDP
17,Sudan,272.0
152,Japan,236.7
109,Singapore,173.1
120,Venezuela,164.3
117,Lebanon,164.1
...,...,...
9,Afghanistan,8.3
161,Cayman Islands,7.6
65,Turkmenistan,4.6
160,Kuwait,3.0


In [64]:
# relationship between Gov. Budget and Debt/GDP
df[['name','Gov. Budget', 'Debt/GDP']]

,name,Gov. Budget,Debt/GDP
0,Tunisia,-6.600,79.80
1,Vietnam,-4.000,32.90
2,Ecuador,-1.900,50.60
3,Puerto Rico,-3.145,59.35
4,United States,-6.400,124.30
...,...,...,...
168,New Caledonia,-3.250,48.15
169,Kosovo,0.200,16.90
170,Malaysia,-4.100,70.40
171,Fiji,-3.400,78.30


In [50]:
df[['name','Debt/GDP']].sort_values(by = 'Debt/GDP',ascending = False)

,name,Debt/GDP
17,Sudan,272.0
152,Japan,236.7
109,Singapore,173.1
120,Venezuela,164.3
117,Lebanon,164.1
...,...,...
9,Afghanistan,8.3
161,Cayman Islands,7.6
65,Turkmenistan,4.6
160,Kuwait,3.0


In [58]:
# Current Account column
df[['Current Account']].sort_values(by = 'Current Account',ascending = False)

,Current Account
166,31.4
160,28.2
48,24.6
130,18.5
109,17.5
...,...
54,-20.4
137,-21.1
12,-21.7
28,-25.2


In [59]:
# country has max cuurent account
df.loc[df['Current Account'] == 31.4]

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
166,Macau,Macanese pataca,UnKnown,Portuguese,22.166667,113.55,30.0,Asia,Eastern Asia,['CHN'],50,1.45,4.75,0.12,1.9,-3.0,38.8,31.4,0.69


In [60]:
# country has min cuurent account
df.loc[df['Current Account'] == -31.5]

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
117,Lebanon,Lebanese pound,Beirut,Arabic,33.833333,35.833333,10452.0,Asia,Western Asia,"['ISR', 'SYR']",20,1.45,20.0,14.3,11.5,-0.4,164.1,-31.5,5.8


In [54]:
df.head()

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
0,Tunisia,Tunisian dinar,Tunis,Arabic,34.000000,9.000000,163610.0,Africa,Northern Africa,"['DZA', 'LBY']",53,1.80,7.50,5.30,15.70,-6.600,79.80,-2.7,11.89
1,Vietnam,Vietnamese đồng,Hanoi,Vietnamese,16.166667,107.833333,331212.0,Asia,South-Eastern Asia,"['KHM', 'CHN', 'LAO']",476,6.88,4.50,3.19,2.24,-4.000,32.90,6.1,101.30
2,Ecuador,United States dollar,Quito,Spanish,-2.000000,-77.500000,276841.0,Americas,South America,"['COL', 'PER']",125,3.46,8.50,0.72,4.50,-1.900,50.60,2.8,17.97
3,Puerto Rico,United States dollar,San Juan,English,18.250000,-66.500000,8870.0,Americas,Caribbean,[],126,0.60,5.75,1.20,5.70,-3.145,59.35,-0.9,3.20
4,United States,United States dollar,"Washington, D.C.",English,38.000000,-97.000000,9372610.0,Americas,North America,"['CAN', 'MEX']",29185,3.00,4.50,2.70,4.20,-6.400,124.30,-3.9,341.15


In [55]:
# Population column
df.groupby('region')['Population'].mean()

region
Africa      27.399800
Americas    36.093793
Asia        97.610208
Europe      17.977073
Oceania      8.912000
Name: Population, dtype: float64

In [56]:
# top 5 countries in Population
df[['name','languages','area', 'GDP Growth', 'Population']].sort_values(by = 'Population', ascending = False).head()

,name,languages,area,GDP Growth,Population
108,China,Chinese,9706961.0,1.10,1408.00
49,India,English,3287590.0,2.00,1398.60
4,United States,English,9372610.0,3.00,341.15
94,Indonesia,Indonesian,1904569.0,4.04,281.60
132,Pakistan,English,881912.0,5.79,251.27


In [72]:
# outlier in Population
Q3 = 34.430000
IQR = 34.430000 - 3.540000
up = Q3 + 1.5 *IQR
df[df['Population'] > up].sort_values(by = 'Population', ascending = False)

,name,currency,capital,languages,latitude,longitude,area,region,subregion,borders,GDP,GDP Growth,Interest Rate,Inflation Rate,Jobless Rate,Gov. Budget,Debt/GDP,Current Account,Population
108,China,Chinese yuan,Beijing,Chinese,35.000000,105.000000,9706961.0,Asia,Eastern Asia,"['AFG', 'BTN', 'MMR', 'HKG', 'IND', 'KAZ', 'NP...",18744,1.10,3.00,0.00,5.20,-6.5,88.30,2.20,1408.00
49,India,Indian rupee,New Delhi,English,20.000000,77.000000,3287590.0,Asia,Southern Asia,"['BGD', 'BTN', 'MMR', 'CHN', 'NPL', 'PAK']",3913,2.00,5.50,1.55,5.20,-4.8,81.59,-0.60,1398.60
4,United States,United States dollar,"Washington, D.C.",English,38.000000,-97.000000,9372610.0,Americas,North America,"['CAN', 'MEX']",29185,3.00,4.50,2.70,4.20,-6.4,124.30,-3.90,341.15
94,Indonesia,Indonesian rupiah,Jakarta,Indonesian,-5.000000,120.000000,1904569.0,Asia,South-Eastern Asia,"['TLS', 'MYS', 'PNG']",1396,4.04,5.00,2.37,4.76,-2.3,38.80,-0.63,281.60
132,Pakistan,Pakistani rupee,Islamabad,English,30.000000,70.000000,881912.0,Asia,Southern Asia,"['AFG', 'CHN', 'IND', 'IRN']",373,5.79,11.00,4.10,5.50,-6.8,80.00,-0.50,251.27
104,Nigeria,Nigerian naira,Abuja,English,10.000000,8.000000,923768.0,Africa,Western Africa,"['BEN', 'CMR', 'TCD', 'NER']",188,10.00,27.50,21.88,4.30,-6.1,52.90,-0.50,232.68
86,Brazil,Brazilian real,Brasília,Portuguese,-10.000000,-55.000000,8515767.0,Americas,South America,"['ARG', 'BOL', 'COL', 'GUF', 'GUY', 'PRY', 'PE...",2179,1.40,15.00,5.23,5.80,-8.5,76.50,-2.55,212.58
77,Bangladesh,Bangladeshi taka,Dhaka,Bengali,24.000000,90.000000,147570.0,Asia,Southern Asia,"['MMR', 'IND']",450,5.82,10.00,8.55,4.70,-4.6,31.70,-1.40,171.60
14,Russia,Russian ruble,Moscow,Russian,60.000000,100.000000,17098242.0,Europe,Eastern Europe,"['AZE', 'BLR', 'CHN', 'EST', 'FIN', 'GEO', 'KA...",2174,-0.80,18.00,8.80,2.20,-1.7,16.40,2.90,146.20
100,Ethiopia,Ethiopian birr,Addis Ababa,Amharic,8.000000,38.000000,1104300.0,Africa,Eastern Africa,"['DJI', 'ERI', 'KEN', 'SOM', 'SSD', 'SDN']",164,1.25,15.00,13.70,18.90,-2.0,32.00,-3.40,132.06


World Economics Data Summary

Dataset Overview:
. Contains 173 countries with 19 features (name, currency, GDP, population,...)
. Data includes geographic (region, borders) and economic metrics (GDP, inflation, debt, ...)
. Some columns (e.g., GDP Growth, Interest Rate) have missing values for smaller countries

Geographic Insights:
. Regions: Africa almost contains 50 countries, Asia, Americas, Europe, Oceania (least countries)
. English is the most common language, followed by Arabic, French, and Spanish
. Largest countries by area: Russia (17M km²), Canada (9.9M km²) ,China (9.7M km²), United States (9.4M km²)

Economic Insights:
. United States has highest GDP followed by China then Germany
. Sudan (272%) highest Debt/GDP ratios, followed by Japan (236.7%), Singapore (173.1%)
. Africa has the highest average GDP Growth among regions, with countries like Algeria (18.6%) standing out
. High Jobless rates in Africa (South Africa 33.2%) and some Asian subregions (West Asia 28.6%)

Financial Metrics:
. Inflation varies widely: Venezuela (172%) and Sudan (146.6%) have extreme rates, while Fiji (-0.4%) shows deflation
. Most countries have deficits (negative Gov. Budget) like United States (-6.4%), Tunisia (-6.6%)
. Current Account: Macau (31.4%) and Kuwait (28.2%) have strong surpluses, Lebanon (-31.5%) has a large deficit

Population Insights:
. Asia has maximum avg population (97.6M), Americas (36M), Africa (27.4M), Europe (18M), Oceania (8.9M)
. Top 5 populous countries: China, India , United States, Indonesia, Pakistan
. China and India greatly influence population distribution (Population outliers)
